flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/HsMemoryDataManager.java

```diff
@@ -0,0 +1,673 @@
+        private final Deque<BufferContext> consumeBuffers = new LinkedList<>();
+
+        @GuardedBy("subpartitionLock")
+        private final Map<Integer, BufferContext> bufferIndexToContexts;
+
+        private final int targetChannel;
+
+        HsSubpartitionDataBuffer(int targetChannel) {
+            this.targetChannel = targetChannel;
+            this.bufferIndexToContexts = new HashMap<>();
+        }
+
+        // ------------------------------------------------------------------------
+        //  Called by Consumer
+        // ------------------------------------------------------------------------
+
+        public Buffer.DataType getFirstDataType(int expectedBufferIndex) {
+            synchronized (subpartitionLock) {
+                while (!consumeBuffers.isEmpty() && consumeBuffers.peekFirst().isReleased) {
+                    consumeBuffers.pollFirst();
+                }
+                if (consumeBuffers.isEmpty()
+                        || consumeBuffers.peekFirst().bufferIndexAndChannel.getBufferIndex()
+                                != expectedBufferIndex) {
+                    return Buffer.DataType.NONE;
+                }
+                BufferContext bufferContext = checkNotNull(consumeBuffers.peekFirst());
+                return bufferContext.buffer.getDataType();
+            }
+        }
+
+        public BufferAndNextDataType pollBuffer(int expectedBufferIndex) {
+            BufferContext bufferContext;
+            Buffer.DataType nextDataType;
+            try {
+                lock.readLock().lock();
+                synchronized (subpartitionLock) {
+                    while (!consumeBuffers.isEmpty() && consumeBuffers.peekFirst().isReleased) {
+                        consumeBuffers.pollFirst();
+                    }
+                    if (consumeBuffers.isEmpty()
+                            || consumeBuffers.peekFirst().bufferIndexAndChannel.getBufferIndex()
+                                    != expectedBufferIndex) {
+                        return null;
+                    }
+
+                    bufferContext = checkNotNull(consumeBuffers.pollFirst());
+
+                    bufferContext.onBufferConsumed();
+                    while (!consumeBuffers.isEmpty() && consumeBuffers.peekFirst().isReleased) {
+                        consumeBuffers.pollFirst();
+                    }
+                    BufferContext next = consumeBuffers.peekFirst();
+                    nextDataType = next == null ? Buffer.DataType.NONE : next.buffer.getDataType();
+                }
+            } finally {
+                lock.readLock().unlock();
+            }
+            Optional<Decision> decisionOpt =
+                    spillStrategy.onBufferConsumed(bufferContext.bufferIndexAndChannel);
+            // Calling handleDecision must be outside the read lock, because write lock may be
+            // accessed inside, which may block forever.
+            handleDecision(decisionOpt);
```

Comment:
It might be better to always deal with the strategy and decision in the memory data manager.

Reply:
done.